// Partida/ComprobarGanador.js
var connection = require("Conexion");

module.exports = ComprobarGanador;

var leer = require("Tablero");
var card = require("Cartas");

function ComprobarGanador(partida, callback) {
    leer.Lectura.tablero(partida, function (err, tablero) {
        if (!err) {
            ComprobarGanadorMineros(partida, tablero, function (err, resultado) {
                if (!err) {
                    if (resultado) {
                        callback(null, "Buscador");
                    } else {

                        var tableroVisto = new Array(7);
                        for (var x = 0; x < 7; x++) {
                            tableroVisto[x] = new Array(7);
                        }

                        ComprobarGanadorSaboteador(tablero, tableroVisto, "", 3, 0, function (err, resultado) {
                            if (!err) {
                                if (resultado) {
                                    callback(null, "Saboteador");
                                } else {
                                    callback(null, "No Ganador");
                                }
                            } else {
                                callback(err);
                            }
                        });
                    }
                } else {
                    callback(err);
                }
            });
        } else {

        }
    });
}

function ComprobarGanadorSaboteador(tablero, tableroVisto, procedencia, fila, columna, callback) {
    var casilla = tablero[fila][columna].Tipo;
    tableroVisto[fila][columna] = true;
    var gana = null;
    var f = Number(fila);
    var c = Number(columna);

    if (casilla.length === 1) {
        callback(null, true, tableroVisto);
    } else {
        if (casilla === "Inicio" || (casilla.indexOf("u") !== -1)) {
            if (f - 1 >= 0) {
                if (tablero[f - 1][c] !== undefined){
                    if (!tableroVisto[f - 1][c]) {
                        ComprobarGanadorSaboteador(tablero, tableroVisto, "d", f - 1, c, function (err, resultadoArriba, tab) {
                            if (!err) {
                                if (resultadoArriba && gana !== false) {
                                    gana = true;
                                } else {
                                    tableroVisto = tab;
                                    gana = false;
                                }
                            } else {
                                callback(err);
                            }
                        });
                    } else {
                        if (gana !== false) {
                            gana = true;
                        }
                    }
                } else{
                    gana = false;
                }
            } else {
                if (gana !== false) {
                    gana = true;
                }
            }
        }
        if (casilla === "Inicio" || (casilla.indexOf("d") !== -1)) {
            if (f + 1 <= 6) {
                if (tablero[f + 1][c] !== undefined){
                    if (!tableroVisto[f + 1][c]) {
                        ComprobarGanadorSaboteador(tablero, tableroVisto, "u", f + 1, c, function (err, resultadoAbajo, tab) {
                            if (!err) {
                                if (resultadoAbajo && gana !== false) {
                                    gana = true;
                                } else {
                                    tableroVisto = tab;
                                    gana = false;
                                }
                            } else {
                                callback(err);
                            }
                        });
                    } else {
                        if (gana !== false) {
                            gana = true;
                        }
                    }
                } else{
                    gana = false;
                }
            } else {
                if (gana !== false) {
                    gana = true;
                }
            }
        }
        if (casilla === "Inicio" || casilla.indexOf("r") !== -1) {
            if (c + 1 <= 6) {
                if (tablero[f][c + 1] !== undefined) {
                    if (!tableroVisto[f][c + 1]) {
                        ComprobarGanadorSaboteador(tablero, tableroVisto, "l", f, c + 1, function (err, resultadoDerecha, tab) {
                            if (!err) {
                                if (resultadoDerecha && gana !== false) {
                                    gana = true;
                                } else {
                                    tableroVisto = tab;
                                    gana = false;
                                }
                            } else {
                                callback(err);
                            }
                        });
                    } else {
                        if (gana !== false) {
                            gana = true;
                        }
                    }
                } else{
                    gana = false;
                }
            } else {
                if (gana !== false) {
                    gana = true;
                }
            }
        }
        if (casilla === "Inicio" || casilla.indexOf("l") !== -1) {
            if (c - 1 >= 0) {
                if (tablero[f][c - 1] !== undefined) {
                    if (!tableroVisto[f][c - 1]) {
                        ComprobarGanadorSaboteador(tablero, tableroVisto, "r", f, c - 1, function (err, resultadoIzquierda, tab) {
                            if (!err) {
                                if (resultadoIzquierda && gana !== false) {
                                    gana = true;
                                } else {
                                    tableroVisto = tab;
                                    gana = false;
                                }
                            } else {
                                callback(err);
                            }
                        });
                    } else {
                        if (gana !== false) {
                            gana = true;
                        }
                    }
                } else{
                    gana = false;
                }
            } else {
                if (gana !== false) {
                    gana = true;
                }
            }
        }
        callback(null, gana, tableroVisto);
    }
}

function ComprobarGanadorMineros(partida, tablero, callback) {

    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }
    conexion.connect(function (err) {
        if (!err) {
            var sql = "select ID from partidas where partidas.Nombre = '" + partida + "';";

            conexion.query(sql, function (err, resultado) {
                if (!err) {
                    var ID_partida = resultado[0].ID;

                    sql = "Select Fila, Columna " +
                            "from poner " +
                            "where ID_Partida = '" + ID_partida + "' and ID_Carta = 2;";

                    conexion.query(sql, function (err, oro) {
                        if (!err) {
                            fichaArriba(tablero[oro[0].Fila - 1][oro[0].Columna], function (err, resultado) {
                                if (!err) {
                                    if (resultado) {
                                        callback(null, true);
                                        conexion.end();
                                    } else {
                                        fichaIzquierda(tablero[oro[0].Fila][oro[0].Columna - 1], function (err, resultado) {
                                            if (!err) {
                                                if (resultado) {
                                                    callback(null, true);
                                                    conexion.end();
                                                } else {
                                                    fichaAbajo(tablero[oro[0].Fila + 1][oro[0].Columna], function (err, resultado) {
                                                        if (!err) {
                                                            if (resultado) {
                                                                callback(null, true);
                                                            } else {
                                                                callback(null, false);
                                                            }
                                                        } else {
                                                            callback(err);
                                                        }
                                                        conexion.end();
                                                    });
                                                }
                                            }
                                        });
                                    }
                                } else {
                                    callback(err);
                                }
                            });
                        } else {
                            callback(err);
                        }
                    });
                } else {
                    callback(err);
                }
            });
        } else {
            callback(err);
        }
    });
}

function fichaArriba(fichaArriba, callback) {
    if (fichaArriba !== undefined) {
        callback(null, true);
    } else {
        callback(null, false);
    }
}

function fichaAbajo(fichaAbajo, callback) {
    if (fichaAbajo !== undefined) {
        callback(null, true);
    } else {
        callback(null, false);
    }
}

function fichaIzquierda(fichaIzquierda, callback) {
    if (fichaIzquierda !== undefined) {
        callback(null, true);
    } else {
        callback(null, false);
    }
}
