// Partida/LeerPartidas.js
var connection = require("Conexion");

module.exports = {
    Abiertas: LeerPartidasAbiertas,
    AbiertasPrivadas: LeerPartidasAbiertasPrivadas,
    ActivasPrivadas: LeerPartidasActivasPrivadas,
    CerradasPrivadas: LeerPartidasCerradasPrivadas,
    PartidaActual: LeerPartidaActual
};

//Funcion que lee todas las partidas abiertas de cualquier usuario
function LeerPartidasAbiertas(usuario, callback) {
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }
    conexion.connect(function (err) {
        if (err) {
            callback(err);
        } else {
            var sql = "select ID from usuario where Nombre_Usuario = '" + usuario + "';";
            conexion.query(sql, function (err, usuario) {
                if (!err) {
                    var IDUsuario = usuario[0].ID;
                    sql = "select Nombre, Creador, Estado, Numero_Jugadores, GROUP_CONCAT(DISTINCT usuario.Nombre_Usuario SEPARATOR ', ') as JugadoresUnidos " +
                            "from partidas, juega, usuario " +
                            "where partidas.Estado = 'Abierta' and partidas.ID = juega.ID_Partida AND juega.ID_Usuario = usuario.ID and juega.ID_Partida NOT IN (select juega.ID_Partida from juega where juega.ID_Usuario = '" + IDUsuario + "') " +
                            "GROUP BY juega.ID_Partida;";
                    conexion.query(sql, function (err, filas) {
                        if (err) {
                            callback(err);
                        } else {
                            callback(null, filas);
                        }
                        conexion.end();
                    });
                } else {
                    callback(err);
                }
            });

        }
    });
}

//Funcion que lee las partidas abiertas que ha creado un usuario
function LeerPartidasAbiertasPrivadas(usuario, callback) {
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }
    conexion.connect(function (err) {
        if (err) {
            callback(err);
        } else {
            var sql = "select Nombre, Fecha, Numero_Jugadores, Turno " +
                    "from partidas" +
                    " where partidas.Estado = 'Abierta' and partidas.Creador = '" + usuario + "';";
            conexion.query(sql, function (err, filas) {
                if (err) {
                    callback(err);
                } else {
                    callback(null, filas);
                }
                conexion.end();
            });
        }
    });
}
;

//Funcion que lee las partidas activas creadas por el usuario
function LeerPartidasActivasPrivadas(usuario, callback) {
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }
    conexion.connect(function (err) {
        if (!err) {
            var sql = "select ID from usuario where Nombre_Usuario = '" + usuario + "';";
            conexion.query(sql, function (err, usuario) {
                if (!err) {
                    var IDUsuario = usuario[0].ID;
                    var sql = "select distinct Nombre, Creador, Turno, Fecha " +
                            "from partidas, juega " +
                            "where partidas.Estado = 'Activa' and juega.ID_Partida IN (select juega.ID_Partida from juega where juega.ID_Usuario = '" + IDUsuario + "') ";
                    conexion.query(sql, function (err, filas) {
                        if (err) {
                            callback(err);
                        } else {
                            callback(null, filas);
                        }
                        conexion.end();
                    });
                } else {
                    callback(err);
                }
            });
        } else {
            callback(err);
        }
    });
}
;

//Funcion que lee las partidas termianadas por el usuario ya Cerradas
function LeerPartidasCerradasPrivadas(usuario, callback) {
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }
    conexion.connect(function (err) {
        if (!err) {
            var sql = "select ID from usuario where Nombre_Usuario = '" + usuario + "';";
            conexion.query(sql, function (err, usuario) {
                if (!err) {
                    var IDUsuario = usuario[0].ID;
                    var sql = "select distinct Nombre, Creador, Estado, Numero_Jugadores, Ganador " +
                              "from partidas, juega " +
                              "where partidas.Estado = 'Cerrada'  and juega.ID_Partida IN (select juega.ID_Partida from juega where juega.ID_Usuario = '" + IDUsuario + "'); ";
                    conexion.query(sql, function (err, filas) {
                        if (err) {
                            callback(err);
                        } else {
                            callback(null, filas);
                        }
                        conexion.end();
                    });
                } else {
                    callback(err);
                }
            });
        } else {
            callback(err);
        }
    });
}

//Funcion que lee la partida que se esta jugando
function LeerPartidaActual(nombre, callback) {
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }
    conexion.connect(function (err) {
        if (err) {
            callback(err);
        } else {
            var sql = "select partidas.ID as game, usuario.ID as user, Turno, Creador, Turnos_restantes, Numero_Jugadores, usuario.Nombre_Usuario, usuario.Foto, juega.Rol, Ganador " +
                    "from partidas, juega, usuario " +
                    "where partidas.Nombre = '" + nombre + "' and partidas.ID = juega.ID_Partida AND juega.ID_Usuario = usuario.ID;";
            conexion.query(sql, function (err, filas) {
                if (err) {
                    callback(err);
                } else {
                    callback(null, filas);
                }
            });
        }
        conexion.end();
    });
}
;
