// Partida/LeerPartidas.js
var connection = require("Conexion");

module.exports = {
    Abiertas: LeerPartidasAbiertas,
    AbiertasPrivadas: LeerPartidasAbiertasPrivadas,
    ActivasPrivadas: LeerPartidasActivasPrivadas,
    CerradasPrivadas: LeerPartidasCerradasPrivadas
};

function LeerPartidasAbiertas(usuario, callback) {
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }
    conexion.connect(function (err) {
        if (err) {
            callback(err);
        } else {
            var sql = "select * " +
                    "from partidas" + 
                    " where partidas.Estado = 'Abierta' and partidas.Creador != '" + usuario + "';";
            conexion.query(sql, function (err, filas) {
                if (err) {
                    callback(err);
                } else {
                    callback(null, filas);
                }
            });
        }
        conexion.end();
    });
}

function LeerPartidasAbiertasPrivadas(usuario, callback) {
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }
    conexion.connect(function (err) {
        if (err) {
            callback(err);
        } else {
            var sql = "select Nombre, Fecha, Numero_Jugadores, Turno " +
                    "from partidas" +
                    " where partidas.Estado = 'Abierta' and partidas.Creador = '" + usuario + "';";
            conexion.query(sql, function (err, filas) {
                if (err) {
                    callback(err);
                } else {
                    callback(null, filas);
                }
            });
        }
        conexion.end();
    });
};

function LeerPartidasActivasPrivadas(usuario, callback){
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }
    conexion.connect(function (err) {
        if (err) {
            callback(err);
        } else {
            var sql = "select Nombre, Creador, Turno, Fecha " +
                    "from partidas" +
                    " where partidas.Estado = 'Activa';";
            conexion.query(sql, function (err, filas) {
                if (err) {
                    callback(err);
                } else {
                    callback(null, filas);
                }
            });
        }
        conexion.end();
    });
};

function LeerPartidasCerradasPrivadas(usuario, callback){
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }
    conexion.connect(function (err) {
        if (err) {
            callback(err);
        } else {
            var sql = "select Nombre, Creador, Estado, Numero_Jugadores " +
                    "from partidas" +
                    " where partidas.Estado = 'Cerrada';";
            conexion.query(sql, function (err, filas) {
                if (err) {
                    callback(err);
                } else {
                    callback(null, filas);
                }
            });
        }
        conexion.end();
    });
};
