// Tablero/leerTablero.js
var connection = require("Conexion");

module.exports = {
    tablero: leerTablero,
    casilla: LeerDatosCasilla
};

function leerTablero(partida, callback){
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }

    conexion.connect(function (err) {
        if (!err) {
            var sql = "select ID from partidas where partidas.Nombre ='" + partida + "';";

            conexion.query(sql, function (err, resultado) {
                if (!err) {
                    var ID_partida = resultado[0].ID;
                    sql = "select poner.Fila, poner.Columna, poner.ID_Carta, cartas.URL, cartas.Tipo, poner.Descubierta " +
                            "from poner, cartas " +
                            "where poner.ID_Partida = '" + ID_partida + "' and poner.ID_Carta = cartas.ID;";
                    conexion.query(sql, function (err, resultado) {
                        if (!err) {
                            var matriz = new Array(7);
                            for (var x = 0; x < 7; x++) {
                                matriz[x] = new Array(7);
                            }

                            resultado.forEach(function (ficha) {
                                var aux = {
                                    ID_Carta: ficha.ID_Carta,
                                    Tipo: ficha.Tipo,
                                    URL: ficha.URL,
                                    Descubierta: ficha.Descubierta
                                };

                                matriz[ficha.Fila][ficha.Columna] = aux;

                            });

                            callback(null, matriz);
                        } else {
                            callback(err);
                        }
                        conexion.end();
                    });
                } else {
                    callback(err);
                }
            });
        } else {
            callback(err);
        }


    });
}

function LeerDatosCasilla(ID, partida,fila,columna, callback) {
    var conexion = connection.getConexion();
    if (callback === undefined) {
        callback = function () {};
    }

    conexion.connect(function (err) {
        if (!err) {
            var sql = "select ID from partidas where partidas.Nombre ='" + partida + "';";

            conexion.query(sql, function (err, resultado) {
                if (!err) {
                    var ID_partida = resultado[0].ID;
                   
                    sql = "select usuario.Nombre_Usuario " +
                            "from poner, usuario " +
                            "where poner.ID_Usuario = usuario.ID and poner.ID_Carta = '" + ID + 
                            "' and poner.ID_Partida = '" + ID_partida + 
                            "' and poner.Fila = '" +fila +"' and poner.Columna= '"+columna+"';";
                    
                    conexion.query(sql, function (err, resultado) {
                        if (!err) {
                            callback(null,resultado);
                        } else {
                            callback(err);
                        }

                    });
                } else {
                    callback(err);
                }
            });
        } else {
            callback(err);
        }
    });
}
