// Tablero/ponerFichaTablero.js
var connection = require("Conexion");

module.exports = PonerFichaTablero;

var tab = require("./leerTablero.js");

function PonerFichaTablero(usuario, usuarioAfectado, partida, carta, fila, columna, callback) {
    if (callback === undefined) {
        callback = function () {};
    }
    var conexion = connection.getConexion();

    conexion.connect(function (err) {
        if (!err) {
            var sql = "SELECT Tipo FROM cartas WHERE cartas.ID = ?;";

            conexion.query(sql, [carta], function (err, Tipocarta) {
                if (!err) {
                    var tipo = Tipocarta[0].Tipo;

                    sql = "select ID from partidas where partidas.Nombre = ?;";
                    conexion.query(sql, [partida], function (err, IDPart) {
                        if (!err) {
                            var ID_partida = IDPart[0].ID;
                            sql = "select ID from usuario where usuario.Nombre_Usuario = ?;";
                            conexion.query(sql, [usuario], function (err, IDUsua) {
                                if (!err) {
                                    var ID_usuario = IDUsua[0].ID;
                                    tab.tablero(partida, function (err, tablero) {
                                        if (!err) {

                                            var aux = {
                                                ID_Carta: Number(carta),
                                                Tipo: tipo,
                                                URL: "null"
                                            };
                                            switch (tipo) {
                                                case "DestruirPico":
                                                    ponerFichaPico(ID_partida, usuarioAfectado, 1, function (err, resultado) {
                                                        if (!err) {
                                                            callback(null, resultado);
                                                        } else {
                                                            callback(err);
                                                        }
                                                    });
                                                    break;
                                                case "ArreglarPico":
                                                    ponerFichaPico(ID_partida, usuarioAfectado, 0, function (err, resultado) {
                                                        if (!err) {
                                                            callback(null, resultado);
                                                        } else {
                                                            callback(err);
                                                        }
                                                    });
                                                    break;
                                                case "Lupa":
                                                    ponerFichaLupa(ID_partida, tablero, fila, columna, function (err, resultado) {
                                                        if (!err) {
                                                            callback(null, resultado);
                                                        } else {
                                                            callback(err);
                                                        }
                                                    });
                                                    break;
                                                case "Bomba":
                                                    ponerFichaBomba(ID_partida, tablero, fila, columna, function (err, resultado) {
                                                        if (!err) {
                                                            callback(null, resultado);
                                                        } else {
                                                            callback(err);
                                                        }
                                                    });
                                                    break;
                                                default:
                                                    ponerFichaNormal(ID_partida, ID_usuario, tablero, aux, fila, columna, function (err, resultado) {
                                                        if (!err) {
                                                            callback(null, resultado);
                                                        } else {
                                                            callback(err);
                                                        }
                                                    });
                                                    break;
                                            }

                                        } else {
                                            callback(err);
                                        }
                                        conexion.end();
                                    });
                                } else {
                                    callback(err);
                                }
                            });
                        } else {
                            callback(err);
                        }
                    });
                } else {
                    callback(err);
                }
            });
        } else {
            callback(err);
        }
    });
}

function ponerFichaPico(ID_partida, usuarioAfectado, pico, callback) {
    if (callback === undefined) {
        callback = function () {};
    }
    var conexion = connection.getConexion();

    conexion.connect(function (err) {
        if (!err) {
            var sql = "Update juega " +
                    "set Pico = ? " +
                    "where ID_Partida = ? and ID_Usuario = ?;";
            conexion.query(sql, [pico, ID_partida, usuarioAfectado], function (err, resultado) {
                if (!err) {
                    callback(null, true);
                } else {
                    callback(err);
                }
                conexion.end();
            });
        } else {
            callback(err);
        }
    });
}

function ponerFichaLupa(ID_partida, tablero, fila, columna, callback) {
    if (callback === undefined) {
        callback = function () {};
    }

    if (tablero[fila][columna].Tipo === "Oro" || tablero[fila][columna].Tipo === "Fracaso") {
        var conexion = connection.getConexion();

        var casilla = [];
        casilla.push({fila: fila, columna: columna});

        conexion.connect(function (err) {
            if (!err) {
                descubrirFicha(casilla, ID_partida, function (err, resultado) {
                    if (!err) {
                        callback(null, true);
                    } else {
                        callback(err);
                    }
                });
            } else {
                callback(err);
            }
        });
    } else {
        callback(null, false);
    }
}

function ponerFichaBomba(ID_partida, tablero, fila, columna, callback) {
    if (callback === undefined) {
        callback = function () {};
    }
    var conexion = connection.getConexion();

    if (tablero[fila][columna] !== undefined && tablero[fila][columna].ID_Carta > 4) {
        conexion.connect(function (err) {
            if (!err) {
                var sql = "delete " +
                        "from poner " +
                        "where ID_Partida = ? and Fila = ? and Columna = ?;";
                conexion.query(sql, [ID_partida, fila, columna], function (err, resultado) {
                    if (!err) {
                        callback(null, true);
                    } else {
                        callback(err);
                    }
                    conexion.end();
                });
            } else {
                callback(err);
            }
        });
    } else {
        callback(null, false);
    }
}

function ponerFichaNormal(ID_partida, ID_usuario, tablero, aux, fila, columna, callback) {

    tablero[fila][columna] = aux;

    var tableroVisto = new Array(7);

    for (var x = 0; x < 7; x++) {
        tableroVisto[x] = new Array(7);
    }
    var conexion = connection.getConexion();

    conexion.connect(function (err) {
        if (!err) {
            comprobarPosicionFicha(tablero, tableroVisto, fila, columna, function (err, LlegaAInicio) {
                if (!err) {
                    comprobarValidezFicha(tablero, fila, columna, function (valida, CasillasDescubrir) {
                        if (LlegaAInicio && valida) {
                            var casillasDesc = CasillasDescubrir;
                            sql = "INSERT INTO `poner` " +
                                    "(ID, ID_Usuario, ID_Partida, ID_Carta, Fila, Columna) " +
                                    "VALUES (null, ?, ?, ?, ?, ?);";
                            conexion.query(sql, [ID_usuario, ID_partida, aux.ID_Carta, fila, columna], function (err, resultado) {
                                if (!err) {
                                    descubrirFicha(casillasDesc, ID_partida, function (err, resultado) {
                                        if (!err) {
                                            callback(null, resultado);
                                        } else {
                                            callback(err);
                                        }
                                        conexion.end();
                                    });
                                } else {
                                    callback(err);
                                }
                            });
                        } else {
                            if (!LlegaAInicio) {
                                callback(null, "No Inicio");
                            } else if (!valida) {
                                callback(null, "Mal Puesta");
                            }
                            conexion.end();
                        }
                    });
                } else {
                    callback(err);
                }
            });
        }
    });
}

function comprobarPosicionFicha(tablero, tableroVisto, fila, columna, callback) {
    var casilla = tablero[fila][columna].Tipo;
    tableroVisto[fila][columna] = true;
    var puede = false;
    var f = Number(fila);
    var c = Number(columna);

    if (casilla === "Inicio") {
        callback(null, true);
    } else {
        if ((casilla.indexOf("u", 0) !== -1 && f - 1 >= 0) || tablero[f][c].ID_Carta < 5) {
            if (tablero[f - 1][c] !== undefined && !tableroVisto[f - 1][c]) {
                if (!puede) {
                    comprobarPosicionFicha(tablero, tableroVisto, f - 1, c, function (err, resultadoArriba, tab) {
                        if (!err) {
                            if (resultadoArriba) {
                                puede = true;
                            } else {
                                tableroVisto = tab;
                            }
                        } else {
                            callback(err);
                        }
                    });
                }
            }
        }
        if ((casilla.indexOf("d", 0) !== -1 && f + 1 <= 6) || tablero[f][c].ID_Carta < 5) {
            if (tablero[f + 1][c] !== undefined && !tableroVisto[f + 1][c]) {
                if (!puede) {
                    comprobarPosicionFicha(tablero, tableroVisto, f + 1, c, function (err, resultadoAbajo, tab) {
                        if (!err) {
                            if (resultadoAbajo) {
                                puede = true;
                            } else {
                                tableroVisto = tab;
                            }
                        } else {
                            callback(err);
                        }
                    });
                }
            }
        }
        if ((casilla.indexOf("r", 0) !== -1 && c + 1 <= 6) || tablero[f][c].ID_Carta < 5) {
            if (tablero[f][c + 1] !== undefined && !tableroVisto[f][c + 1]) {
                if (!puede) {
                    comprobarPosicionFicha(tablero, tableroVisto, f, c + 1, function (err, resultadoDerecha, tab) {
                        if (!err) {
                            if (resultadoDerecha) {
                                puede = true;
                            } else {
                                tableroVisto = tab;
                            }
                        } else {
                            callback(err);
                        }
                    });
                }
            }
        }
        if ((casilla.indexOf("l", 0) !== -1 && c - 1 >= 0) || tablero[f][c].ID_Carta < 5) {
            if (tablero[f][c - 1] !== undefined && !tableroVisto[f][c - 1]) {
                if (!puede) {
                    comprobarPosicionFicha(tablero, tableroVisto, f, c - 1, function (err, resultadoIzquierda, tab) {
                        if (!err) {
                            if (resultadoIzquierda) {
                                puede = true;
                            } else {
                                tableroVisto = tab;
                            }
                        } else {
                            callback(err);
                        }
                    });
                }
            }
        }

        callback(null, puede, tableroVisto);
    }
}

function comprobarValidezFicha(tablero, fila, columna, callback) {
    var casilla = tablero[fila][columna].Tipo;
    var valida = true;
    var f = Number(fila);
    var c = Number(columna);
    var descubrir = new Array();

    if (casilla.indexOf("u", 0) === -1 && f - 1 >= 0) {
        if (tablero[f - 1][c] !== undefined && (tablero[f - 1][c].Tipo.indexOf("d") !== -1 || tablero[f - 1][c].ID_Carta < 5)) {
            valida = false;
        }
    } else if (casilla.indexOf("u", 0) !== -1 && f - 1 >= 0) {
        if (tablero[f - 1][c] !== undefined) {
            if (tablero[f - 1][c].Tipo.indexOf("d") === -1 && tablero[f - 1][c].ID_Carta > 4) {
                valida = false;
            } else if (tablero[f - 1][c].Tipo === "Fracaso") {
                descubrir.push({fila: f - 1, columna: c});
            }
        }
    }

    if (casilla.indexOf("d", 0) === -1 && f + 1 <= 6) {
        if (tablero[f + 1][c] !== undefined && (tablero[f + 1][c].Tipo.indexOf("u") !== -1 || tablero[f + 1][c].ID_Carta < 5)) {
            valida = false;
        }
    } else if (casilla.indexOf("d", 0) !== -1 && f + 1 <= 6) {
        if (tablero[f + 1][c] !== undefined) {
            if (tablero[f + 1][c].Tipo.indexOf("u") === -1 && tablero[f + 1][c].ID_Carta > 4) {
                valida = false;
            } else if (tablero[f + 1][c].Tipo === "Fracaso") {
                descubrir.push({fila: f + 1, columna: c});
            }
        }
    }

    if (casilla.indexOf("r", 0) === -1 && c + 1 <= 6) {
        if (tablero[f][c + 1] !== undefined && (tablero[f][c + 1].Tipo.indexOf("l") !== -1 || tablero[f][c + 1].ID_Carta < 5)) {
            valida = false;
        }
    } else if (casilla.indexOf("r", 0) !== -1 && c + 1 <= 6) {
        if (tablero[f][c + 1] !== undefined) {
            if (tablero[f][c + 1].Tipo.indexOf("l") === -1 && tablero[f][c + 1].ID_Carta > 4) {
                valida = false;
            } else if (tablero[f][c + 1].Tipo === "Fracaso") {
                descubrir.push({fila: f, columna: c + 1});
            }
        }
    }
    if (casilla.indexOf("l", 0) === -1 && c - 1 >= 0) {
        if (tablero[f][c - 1] !== undefined && (tablero[f][c - 1].Tipo.indexOf("r") !== -1 || tablero[f][c - 1].ID_Carta < 5)) {
            valida = false;
        }
    } else if (casilla.indexOf("l", 0) !== -1 && c - 1 >= 0) {
        if (tablero[f][c - 1] !== undefined) {
            if (tablero[f][c - 1].Tipo.indexOf("r") === -1 && tablero[f][c - 1].ID_Carta > 4) {
                valida = false;
            } else if (tablero[f][c - 1].Tipo < 5) {
                descubrir.push({fila: f, columna: c - 1});
            }
        }
    }
    callback(valida, descubrir);
}

function descubrirFicha(casilla, IDPartida, callback) {
    if (casilla.length !== 0) {
        var conexion = connection.getConexion();

        conexion.connect(function (err) {
            if (!err) {
                var sql = "Update poner " +
                        "set Descubierta = '1' " +
                        "where (";
                for (x = 0; x < casilla.length; x++) {
                    sql = sql + "(Fila = '" + casilla[x].fila + "' and Columna = '" + casilla[x].columna + "')";
                    if (x !== casilla.length - 1) {
                        sql = sql + " or ";
                    }
                }

                sql = sql + ") and ID_Partida = '" + IDPartida + "';";

                conexion.query(sql, function (err, resultado) {
                    if (!err) {
                        callback(null, resultado);
                    } else {
                        callback(err);
                    }
                    conexion.end();
                });
            }
        });
    } else {
        callback();
    }
}
